### PR TITLE
Use random_compat

### PR DESCRIPTION
**Pull request by paragonie-scott**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -51,6 +51,7 @@
         }
     ],
     "require": {
+        "paragonie/random_compat": "^1.4",
```
Comment (GrahamCampbell) [inline composer.json]: doing this conflicts with people who want 2.x

Comment (paragonie-scott) [inline composer.json]: Good point. Shall I change it to `^1.4|2.0`?

Comment (GrahamCampbell) [inline composer.json]: `^1.4|^2.0`

**Comment by terrafrost** [comment]
What if we just do away with the whole `version_compare` thing and do this?:

```php
        try {
            return \random_bytes($length);
        } catch (Exception $e) {}

        // at this point we have no choice but to use a pure-PHP CSPRNG
        // ...
```

File: phpseclib/Crypt/Random.php [hunk 2]
```diff
@@ -46,7 +46,7 @@ class Random
      */
     static function string($length)
     {
-        if (version_compare(PHP_VERSION, '7.0.0', '>=')) {
+        if (version_compare(PHP_VERSION, '7.0.0', '>=') || defined('RANDOM_COMPAT_READ_BUFFER')) {
             try {
                 return \random_bytes($length);
             } catch (\Throwable $e) {
```
Comment (GrahamCampbell) [inline phpseclib/Crypt/Random.php]: we need to catch Exception now too

Comment (bantu) [inline phpseclib/Crypt/Random.php]: @GrahamCampbell Exception implements Throwable?

Comment (paragonie-scott) [inline phpseclib/Crypt/Random.php]: Not in PHP 5. I couldn't make Exception implement Throwable with random_compat

Comment (bantu) [inline phpseclib/Crypt/Random.php]: Oh ok. Now I see. Add another empty catch block for Exception?

Comment (GrahamCampbell) [inline phpseclib/Crypt/Random.php]: yeh

**Comment by paragonie-scott** [comment]
OK, how does that look now?

File: phpseclib/Crypt/Random.php [hunk 3]
```diff
@@ -46,64 +46,17 @@ class Random
...
+        try {
+            return \random_bytes($length);
+        } catch (\Throwable $e) {
+            // If a sufficient source of randomness is unavailable, random_bytes() will throw an
+            // object that implements the Throwable interface (Exception, TypeError, Error).
+            // We don't actually need to do anything here. The string() method should just continue
+            // as normal. Note, however, that if we don't have a sufficient source of randomness for
+            // random_bytes(), most of the other calls here will fail too, so we'll end up using
+            // the PHP implementation.
+        } catch (\Exception $e) {
```
Comment (GrahamCampbell) [inline phpseclib/Crypt/Random.php]: It's conventional to catch Exception first, then Throwable.

Comment (paragonie-scott) [inline phpseclib/Crypt/Random.php]: Ok, easily done.

**Comment by paragonie-scott** [comment]
I think the unit tests are failing because of `composer.lock`.

**Comment by terrafrost** [comment]
Looks good! I'll figure out what's up with the unit tests later - I need to go do some errands lol.

Thanks!!

**Comment by terrafrost** [comment]
The composer.lock was the issue and the PR has been merged.

Thanks!

**Pull request merged**